components/submenu_modal/submenu_modal.test.jsx

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React from 'react';
+import {shallow, mount} from 'enzyme';
+import {Modal} from 'react-bootstrap';
+
+import {browserHistory} from 'utils/browser_history';
+
+import SubMenuModal from 'components/submenu_modal/submenu_modal.jsx';
+
+describe('components/submenu_modal', () => {
+    const baseProps = {
+        elements: [
+            {
+                id: 'A',
+                text: 'Text A',
+            },
+            {
+                id: 'B',
+                text: 'Text B',
+                subMenu: [
+                    {
+                        id: 'C',
+                        text: 'Text C',
+                    },
+                ],
+            },
+        ],
+        action: jest.fn(),
+        onHide: jest.fn(),
+        openModal: jest.fn(),
+    };
+
+    test('should match snapshot', () => {
+        const wrapper = shallow(
+            <SubMenuModal {...baseProps}/>
+        );
+        expect(wrapper).toMatchSnapshot();
+    });
+
+    test('should match state when onHide is called', () => {
+        const wrapper = shallow(
+            <SubMenuModal {...baseProps}/>
+        );
+
+        wrapper.setState({show: true});
+        wrapper.instance().onHide();
+        expect(wrapper.state('show')).toEqual(false);
+    });
+
+    test('should have called click function when button is clicked', async () => {
+        browserHistory.push = jest.fn();
+        const action = jest.fn().mockReturnValueOnce();
+        const props = {
+            ...baseProps,
+            action,
+        };
+        const wrapper = mount(
+            <SubMenuModal {...props}/>
+        );
+
+        wrapper.setState({show: true});
+        await wrapper.find('#A').at(1).simulate('click');
+        expect(action).toHaveBeenCalledTimes(1);
+        expect(action).toHaveBeenCalledWith('A');
+        expect(wrapper.state('show')).toEqual(false);
+
+        wrapper.setState({show: true});
+        await wrapper.find('#B').at(1).simulate('click');
+        expect(action).toHaveBeenCalledTimes(2);
+        expect(action).toHaveBeenCalledWith('B');
+        expect(wrapper.state('show')).toEqual(false);
+
+        wrapper.setState({show: true});
+        await wrapper.find('#C').at(1).simulate('click');
+        expect(action).toHaveBeenCalledTimes(4);
```

Comment:
I've seen this and it's correct but didn't understand why called 4 times instead of 3.

Reply:
Great eyes @saturninoabril !  :)
Running `npm run test`, there is no `event.nativeEvent.path` in the click event, to find the original id that triggered the click event.  This path is important as there could possibly be multiple parent nodes that will also receive the click event in the submenu tree.  For built/running code, this isn't an issue because nativeEvent.path has the triggering id in one of the first two elements, and only one onClick is triggered.  

For running tests only, I added the `else if (!event.nativeEvent.path)...` to SubMenuItemAction's onClick handling.  However, each successive parent node will receive onClick events and in the case of this test, 'C' is a submenu and triggers 'B' as well. 

 I am completely open to suggestions on how to do/test this better.